graphstorm-processing/graphstorm_processing/graph_loaders/dist_heterogeneous_loader.py

```diff
@@ -0,0 +1,1560 @@
+            )
+        else:
+            raise RuntimeError(f"Invalid task type: {label_config.task_type}")
+
+    def _create_split_files_from_rates(
+        self,
+        input_df: DataFrame,
+        label_column: str,
+        split_rates: Optional[SplitRates],
+        output_path: str,
+        seed: Optional[int] = None,
+    ) -> Dict:
+        """
+        Given an input dataframe and a list of split rates creates the
+        split masks and writes them to S3 and returns the corresponding
+        metadata.json dict elements.
+
+        Parameters
+        ----------
+        input_df: DataFrame
+            Input dataframe for which we will create split masks.
+        label_column: str
+            The name of the label column. If provided, the values in the column
+            need to be not null for the data point to be included in one of the masks.
+            If an empty string, all rows in the input_df are included in one of train/val/test sets.
+        split_rates: Optional[SplitRates]
+            A SplitRates object indicating the train/val/test split rates.
+            If None, a default split rate of 0.9:0.05:0.05 is used.
+        output_path: str
+            The output path under which we write the masks.
+        seed: int
+            An optional random seed for reproducibility.
+
+        Returns
+        -------
+        The metadata dict elements for the train/test/val masks, to be added to the caller's
+        edge/node type metadata.
+        """
+        # If the user did not provide a split rate we use a default
+        split_metadata = {}
+        if split_rates is None:
+            split_rates = SplitRates(train_rate=0.9, val_rate=0.05, test_rate=0.05)
+        else:
+            if sum(split_rates.tolist()) != 1.0:
+                raise RuntimeError(f"Provided split rates  do not sum to 1: {split_rates}")
+
+        split_list = split_rates.tolist()
+        logging.info(
+            "Creating split files for label column '%s' with split rates: %s",
+            label_column,
+            split_list,
+        )
+
+        rng = default_rng(seed=seed)
+
+        # We use multinomial sampling to create a one-hot
+        # vector indicating train/test/val membership
+        def multinomial_sample(label_col: str) -> Sequence[int]:
+            if label_col in {"", "None", "NaN", None}:
+                return [0, 0, 0]
+            return rng.multinomial(1, split_list).tolist()
+
+        group_col_name = "sample_boolean_mask"  # TODO: Ensure uniqueness of column?
+
+        # TODO: Use PandasUDF and check if it is faster than UDF
+        split_group = F.udf(multinomial_sample, ArrayType(IntegerType()))
+        # Convert label col to string and apply UDF
+        # to create one-hot vector indicating train/test/val membership
+        input_col = F.col(label_column).astype("string") if label_column else F.lit("dummy")
```

Comment:
what does F.lit("dummy") mean?

Reply:
It just creates a fake column with "dummy" as the value everywhere. `lit` here is shorthand for `literal` https://spark.apache.org/docs/3.1.3/api/python/reference/api/pyspark.sql.functions.lit.html
We use this to ensure all rows are included in one of the masks.

The purpose is explained in the docstring:

> If provided, the values in the column need to be not null for the data point to be included in one of the masks. If an empty string, all rows in the input_df are included in one of train/val/test sets.